codec-registration-checker-compiler-plugin/src/main/scala/org/virtuslab/ash/SerializerCheckCompilerPluginComponent.scala

```diff
@@ -108,29 +106,25 @@ class SerializerCheckCompilerPluginComponent(
 
         val foundTypes =
           try {
-            serializerImplDef
-              .collect {
-                case x: Tree if x.tpe != null => x.tpe
-              }
-              .groupBy(_.toString())
-              .filter(_._1.matches(filterRegex))
-              .map(_._2.head)
+            serializerImplDef.collect {
+              case x @ (_: Select | _: GenericApply) if x.tpe != null && x.tpe.toString().matches(filterRegex) => x.tpe
```

Comment:
Okay! so just limiting the types from `Tree` to `Select | GenericApply` shortened the execution time of the compiler phase two-fold right? 🤔

Reply:
Actually... could you elaborate why just these two subtypes were enough? is there a case where a false-positive (false-negative?) compiler error happens because some subtype of `Tree` was NOT covered here?

Reply:
I.e. I'm asking whether we don't sacrifice **soundness for performance** here!

Reply:
> Okay! so just limiting the types from `Tree` to `Select | GenericApply` shortened the execution time of the compiler phase two-fold right? 🤔

Yes. Because there are fewer types for checking nullability and regex and also a list of founded types is used in recursion after and since it's smaller it's faster to process it.

Reply:
> Actually... could you elaborate why just these two subtypes were enough? is there a case where a false-positive (false-negative?) compiler error happens because some subtype of `Tree` was NOT covered here?

I found the minimal set of classes with having successful tests and compilation for the project that uses ASH. Not sure if all possible subtypes of `Tree` are covered.

Reply:
This is not the only time we iterate through the whole tree searching for something, then why is this iteration so expensive?

Converting `Type` to `String`? Or maybe matching regex? Checking for null is rather fast operation.

Reply:
We have iterations on other places but this is the only place where we get all possible types by matching `Tree` which is the root of all types. I agree they are fast but since they are applied on a very big number of types, compilation time is increased for big projects. Since I couldn't optimize checking nullability and matching regexp I tried to decrease the number of classes on which filtering is applied and that made progress in compilation time.

Reply:
Yes, but we iterate through all the classes in all the files, not just one, and it takes 1 second, not 40. So something is wrong with the body of collect, not with the amount of classes we process.

Maybe the way we acquire `string` is wrong, or maybe we can prevent unnecessary regex usage. And if all else fails, there is always parallelisation. 

Either way, work is far from done.